source/model.js

```diff
@@ -91,57 +107,61 @@ function ComputerPlayer(name, symbol, field){
 ComputerPlayer.prototype = Object.create(Player.prototype);
 ComputerPlayer.prototype.constructor = ComputerPlayer;
 
-function Game(humanPlayer, firstPlayerName, secondPlayerName){
+function Iterator(array) {
+    let index = 0;
+    this.getCurrent = function(){
+        return array[index];
+    };
+    this.moveNext = function(){
+        index = index == array.length-1 ? 0 : ++index;
+    }
+}
+
+function Game(players){
     let self = this;
-    let currentPlayerIndex = 0;
+    this.field = singletonField.getInstance();
+    this.iterator = new Iterator(players);
+    let winnerChecker = new WinnerChecker(self.field);
     let gameStatuses = ['Playing', 'Winner', 'Tie'];
-    let field = new Field();
-    let winnerChecker = new WinnerChecker(field);
-    let players = playerCreator();
 
-    this.currentPlayer = players[currentPlayerIndex];
     this.currentStatus = gameStatuses[0];
 
     this.getFieldCells = function(){
-        return field.fieldCells.slice(0);
+        return self.field.fieldCells.slice(0);
     };
 
     this.startTurn = function(){
-        // players[currentPlayerIndex].startMove();
-       self.currentPlayer.startMove();
+        self.iterator.getCurrent().startMove(self);
     };
 
     this.finishTurn = function(cellNum){
-        if (self.currentStatus == gameStatuses[0] && self.currentPlayer.finishMove(cellNum)) {
-            if (winnerChecker.getWinner(cellNum, self.currentPlayer.playerSymbol)) {
+        if (self.currentStatus == gameStatuses[0] && self.iterator.getCurrent().finishMove(cellNum)) {
+            if (winnerChecker.getWinner(cellNum, self.iterator.getCurrent().playerSymbol)) {
                 self.currentStatus = gameStatuses[1];
-            } else if (!field.areCellsAvailable()) {
+            } else if (!self.field.areCellsAvailable()) {
                 self.currentStatus = gameStatuses[2];
             } else {
-                currentPlayerIndex = currentPlayerIndex == 0 ? 1 : 0;
-                self.currentPlayer = players[currentPlayerIndex];
+                self.iterator.moveNext();
                 self.startTurn();
             }
         }
     };
+}
 
-    function playerCreator(){
-        let playersArr = [];
-        playersArr.push(new Player(firstPlayerName, 'X', field));
-
-        let secondPlayer;
-        if (humanPlayer) {
-            secondPlayer = new Player(secondPlayerName, 'O', field);
-        } else {
-            secondPlayer = new ComputerPlayer('Computer', 'O', field);
+function Factory(){
+    this.createPlayer = function(type, name, symbol){
+        let player = '';
+        switch (type) {
```

Comment:
The quality of our factory can be greatly improved. We need to use some dictionary-like array which keys are named after a type of player and values are equal to names of the appropriate objects.
Example:
`var dict = { 'human' : 'Player', 'computer' : 'ComputerPlayer' }`
After we have created the dictionary, we can use the following syntax to create objects:
`player = new window[dict['human']](name, symbol);`
and after this we can replace the "human" string with the incoming parameter 'type'.
This will allow us to get rid of the switch block in the factory and use a single line to instantiate any type of player (including those that don't yet exist).
The final code that should replace the switch is the following:
`player = new window[dict[type]](name, symbol);`
Of course first we have to make sure that such type of player exists.